app/models/targeting.rb

```diff
@@ -36,9 +36,13 @@ def resolve_hosts!
     raise ::Foreman::Exception, _('Cannot resolve hosts without a bookmark or search query') if bookmark.nil? && search_query.blank?
 
     self.search_query = bookmark.query if dynamic? && bookmark.present?
-    self.save!
-    self.touch(:resolved_at)
-    self.hosts = User.as(user.login) { Host.authorized(RESOLVE_PERMISSION, Host).search_for(search_query) }
+    self.resolved_at = Time.zone.now
+    self.validate!
+    # avoid validation of hosts objects - tey will be loaded for no reason.
```

Comment:
afaiu you can define not to validate on the AR relationship definitions, ideally fix it there.